File: C++/0001.Two_Sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

class Solution {
public:
    // function to find indices of two numbers in the array that add up to the target
    std::vector<int> twoSum(const std::vector<int>& nums, int target) {
        // map to store checked numbers and their indices
        std::unordered_map<int, int> checked;
        // size of the input vector
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            // calculate the difference needed to reach the targe
            int diff = target - nums[i];
            // if the difference is found in the map
            if (checked.find(diff) != checked.end()) {
                // return the indices of the two numbers that sum up to the target
                return {checked[diff], i};
            } else {
                // store the current number and its index in the map
                checked[nums[i]] = i;
            }
        }
        // if no solution is found, return an empty vector
        return {};
    }
};

int main() {
    Solution s;
    std::vector<int> result1 = s.twoSum({2, 7, 11, 15}, 9);
    for (int num : result1) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    std::vector<int> result2 = s.twoSum({3, 2, 4}, 6);
    for (int num : result2) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    std::vector<int> result3 = s.twoSum({-1, -2, -3, -4, -5}, -8);
    for (int num : result3) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    return 0;
}

File: Python/0001.Two_Sum.py
class Solution(object):
    def twoSum(self, nums: list[int], target: int) -> list:
        # dictionary to memoize elements already checked along with their indices
        checked = {}
        # length of the list
        n = len(nums)
        # iterate through the list 
        for i in range(n):
            # calculate the difference between the target and the current element
            diff = target - nums[i]
            # check if the difference exists in the dictionary
            if diff in checked:
                # if the difference exists, return the indices of the two elements 
                # that sum up to the target
                return [checked[diff], i]
            else:
                # otherwise, memoize the current element with its index
                checked[nums[i]] = i

if __name__ == '__main__':
    l = Solution()
    print(l.twoSum([2,7,11,15], 9))
    print(l.twoSum([3,2,4], 6))
    print(l.twoSum([-1,-2,-3,-4,-5], -8))